File: spark_submit.scala
spark.stop # stopping the deafult spark repl session to create our own session

import scala.Product
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext._
import org.apache.spark.streaming._
import org.apache.spark.streaming.StreamingContext._
import org.apache.spark.streaming.kafka._

import com.datastax.spark.connector.SomeColumns
import com.datastax.spark.connector.cql.CassandraConnector
import com.datastax.spark.connector.streaming._

val sparkConf = new SparkConf().setAppName("KafkaSparkStreaming").set("spark.cassandra.connection.host","127.0.0.1")
val ssc = new StreamingContext(sparkConf, Seconds(20))
val topicpMap = "mytopic".split(",").map((_, 1.toInt)).toMap
val lines = KafkaUtils.createStream(ssc, "localhost:2181","sparkgroup",topicpMap).map(_._2)
lines.print();
lines.map(line => { val arr = line.split(","); (arr(0),arr(1),arr(2),arr(3)) }).saveToCassandra("sparkdata","cust_data", SomeColumns("fname","lname","url","product","cnt"))
ssc.start
ssc.awaitTermination

File: spark_submit.py
spark.stop # stopping the deafult spark repl session to create our own session

from pyspark.sql import SparkSession
from pyspark.streaming import StreamingContext
from pyspark.sql.functions import explode, split

spark = SparkSession.builder \
    .appName("KafkaSparkStreaming") \
    .config("spark.cassandra.connection.host", "127.0.0.1") \
    .getOrCreate()

# Read from Kafka topic
df = spark \
  .readStream \
  .format("kafka") \
  .option("kafka.bootstrap.servers", "localhost:9092") \
  .option("subscribe", "mytopic") \
  .load()

# Define a function to save RDD to Cassandra
def save_to_cassandra(rdd):
    rdd.toDF() \
       .selectExpr("split(value, ',') as arr") \
       .selectExpr("arr[0] as fname", "arr[1] as lname", "arr[2] as url", "arr[3] as product", "arr[4] as cnt") \
       .write \
       .format("org.apache.spark.sql.cassandra") \
       .options(table="cust_data", keyspace="sparkdata") \
       .mode("append") \
       .save()

# Apply the save_to_cassandra function to the DataFrame
query = df \
    .selectExpr("CAST(value AS STRING)") \
    .writeStream \
    .outputMode("append") \
    .foreachBatch(save_to_cassandra) \
    .start()

query.awaitTermination()

